lib/protoparser/prometheus/proto.go

```diff
@@ -0,0 +1,838 @@
+				return fmt.Errorf("cannot read name")
+			}
+			mf.Name = name
+		case 3:
+			metricType, ok := fc.Int32()
+			if !ok {
+				return fmt.Errorf("cannot read type")
+			}
+			mf.Type = MetricType(metricType)
+		case 4:
+			data, ok := fc.MessageData()
+			if !ok {
+				return fmt.Errorf("cannot read Metric data")
+			}
+			var m Metric
+			if err := m.unmarshalProtobuf(data); err != nil {
+				return fmt.Errorf("cannot unmarshal Metric: %w", err)
+			}
+			mf.Metrics = append(mf.Metrics, &m)
+		}
+	}
+	return nil
+}
+
+// Metric stores Prometheus metric
+//
+//	message Metric {
+//	  repeated LabelPair label        = 1;
+//	  Gauge              gauge        = 2;
+//	  Counter            counter      = 3;
+//	  Summary            summary      = 4;
+//	  Untyped            untyped      = 5;
+//	  Histogram          histogram    = 7;
+//	  int64              timestamp_ms = 6;
+//	}
+//
+// See https://github.com/prometheus/prometheus/blob/aba007148057c1947122b18b2ad606883cc27220/prompb/io/prometheus/client/metrics.proto#L144
+type Metric struct {
+	Labels  []*LabelPair
+	Gauge   *Gauge
+	Counter *Counter
+	Summary *Summary
+	Untyped *Untyped
+	// Histogram *Histogram
+	Timestamp int64
+}
+
+// unmarshalProtobuf decodes src to Metric struct
+func (m *Metric) unmarshalProtobuf(src []byte) (err error) {
+	m.Labels = nil
+	var fc easyproto.FieldContext
+	for len(src) > 0 {
+		src, err = fc.NextField(src)
+		if err != nil {
+			return fmt.Errorf("cannot read next field in Metric message: %w", err)
+		}
+		switch fc.FieldNum {
+		case 1:
+			data, ok := fc.MessageData()
+			if !ok {
+				return fmt.Errorf("cannot read LabelPair data")
+			}
+			var l LabelPair
+			if err := l.unmarshalProtobuf(data); err != nil {
+				return fmt.Errorf("cannot unmarshal Label: %w", err)
+			}
+			m.Labels = append(m.Labels, &l)
+		case 2:
+			data, ok := fc.MessageData()
+			if !ok {
+				return fmt.Errorf("cannot read gauge data")
+			}
+			var gauge Gauge
+			if err := gauge.unmarshalProtobuf(data); err != nil {
+				return fmt.Errorf("cannot unmarshal gauge: %w", err)
+			}
+			m.Gauge = &gauge
+		case 3:
+			data, ok := fc.MessageData()
+			if !ok {
+				return fmt.Errorf("cannot read counter data")
+			}
+			var counter Counter
+			if err := counter.unmarshalProtobuf(data); err != nil {
+				return fmt.Errorf("cannot unmarshal counter: %w", err)
+			}
+			m.Counter = &counter
+		case 4:
+			data, ok := fc.MessageData()
+			if !ok {
+				return fmt.Errorf("cannot read summary data")
+			}
+			var summary Summary
+			if err := summary.unmarshalProtobuf(data); err != nil {
+				return fmt.Errorf("cannot unmarshal summary: %w", err)
+			}
+			m.Summary = &summary
+		case 5:
+			data, ok := fc.MessageData()
+			if !ok {
+				return fmt.Errorf("cannot read untyped data")
+			}
+			var untyped Untyped
+			if err := untyped.unmarshalProtobuf(data); err != nil {
+				return fmt.Errorf("cannot unmarshal Untyped: %w", err)
+			}
+			m.Untyped = &untyped
+		// case 7:
```

Comment:
should this comment be removed?

Reply:
left it for a future, cause support of these types will be added later